### PR TITLE
add unit and integration tests for rbac authorizer

### PR DESCRIPTION
**Pull request by ericchiang**
This PR adds lots of tests for the RBAC authorizer. 

The plan over the next couple days is to add a lot more test cases.

Updates #23396

cc @erictune

### CODE REVIEW

**Comment by ericchiang** [comment]
@erictune this is a bit more fleshed out. Any comments?

**Comment by ericchiang** [comment]
e2e testing woes

```
â€¢ Failure [301.928 seconds]
[k8s.io] Kubelet
/var/lib/jenkins/workspace/node-pull-build-e2e-test@2/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:639
  metrics api
  /var/lib/jenkins/workspace/node-pull-build-e2e-test@2/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:165
    when querying /stats/summary
    /var/lib/jenkins/workspace/node-pull-build-e2e-test@2/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:164
      it should report resource usage through the stats api [It]
      /var/lib/jenkins/workspace/node-pull-build-e2e-test@2/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:163

      Timed out after 300.000s.
      Expected
          <*errors.errorString | 0xc8209f5310>: {
              s: "expected metrics for kubelet",
          }
      to be nil
```

File: test/integration/rbac_test.go [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+						Subjects:   []v1alpha1.Subject{{Kind: "User", Name: "job-writer-namespace"}},
+						RoleRef:    v1.ObjectReference{Kind: "ClusterRole", Name: "write-jobs"},
+					},
+				},
+			},
+			requests: []request{
+				// Create the namespace used later in the test
+				{superUser, "POST", "", "namespaces", "", "", jobNamespace, http.StatusCreated},
+
+				{"user-with-no-permissions", "POST", "batch", "jobs", "job-namespace", "", aJob, http.StatusForbidden},
```
Comment (erictune) [inline test/integration/rbac_test.go]: You are testing denial when the object does not exist.  You should also test denial when the object does exist.

**Comment by erictune** [comment]
Suggest: for each type of object, have a test case where that object is modified, and then authz outcomes change.

**Comment by davidopp** [comment]
Added 1.3 milestone since the associated issue is marked 1.3

**Comment by erictune** [comment]
LGTM

**Comment by erictune** [comment]
lgtm

**Comment by erictune** [comment]
```
--- FAIL: TestRBAC (0.15s)
	rbac_test.go:438: case 0, req 6: pod-reader GET pods expected "200 OK" got "403 Forbidden"
	rbac_test.go:438: case 1, req 7: job-writer GET jobs expected "200 OK" got "403 Forbidden"
	rbac_test.go:438: case 1, req 8: job-writer GET jobs expected "404 Not Found" got "403 Forbidden"
	rbac_test.go:438: case 1, req 9: job-writer POST jobs expected "201 Created" got "403 Forbidden"
	rbac_test.go:438: case 1, req 10: job-writer GET jobs expected "200 OK" got "403 Forbidden"
	rbac_test.go:438: case 1, req 11: job-writer-namespace GET jobs expected "200 OK" got "403 Forbidden"
	rbac_test.go:438: case 1, req 12: job-writer-namespace GET jobs expected "404 Not Found" got "403 Forbidden"
	rbac_test.go:438: case 1, req 13: job-writer-namespace POST jobs expected "201 Created" got "403 Forbidden"
	rbac_test.go:438: case 1, req 14: job-writer-namespace GET jobs expected "200 OK" got "403 Forbidden"
```

**Comment by ericchiang** [comment]
@erictune am looking into this

**Comment by ericchiang** [comment]
git bisect says #27255 is when this started failing.

**Comment by ericchiang** [comment]
@erictune this should be good now

**Comment by erictune** [comment]
LGTM

**Comment by ericchiang** [comment]
@erictune This test failure doesn't look related to the PR

```
# k8s.io/kubernetes/cmd/kubelet
/usr/local/go/pkg/tool/linux_amd64/link: running aarch64-linux-gnu-gcc failed: fork/exec /usr/bin/aarch64-linux-gnu-gcc: cannot allocate memory
```

**Comment by goltermann** [comment]
@k8s-bot e2e test this issue: #IGNORE

**Pull request merged**